test/extended/util/configv1shim.go

```diff
@@ -0,0 +1,1022 @@
+		GroupVersion: groupVersion,
+		APIResources: mergeResources(aList, bList),
+	}
+}
+
+// ConfigClientShim makes sure whenever there's a static
+// manifest present for a config v1 kind, fake client is used
+// instead of the real one.
+type ConfigClientShim struct {
+	configClient configv1client.Interface
+	v1Kinds      map[string]bool
+	fakeClient   *fakeconfigv1client.Clientset
+}
+
+func (c *ConfigClientShim) Discovery() discovery.DiscoveryInterface {
+	return &ConfigV1DiscoveryClientShim{
+		configClient: c.configClient,
+		fakeClient:   c.fakeClient,
+	}
+}
+
+type ConfigV1DiscoveryClientShim struct {
+	configClient configv1client.Interface
+	fakeClient   *fakeconfigv1client.Clientset
+}
+
+func (c *ConfigV1DiscoveryClientShim) ServerGroups() (*metav1.APIGroupList, error) {
+	groups, err := c.configClient.Discovery().ServerGroups()
+	if err != nil {
+		return groups, err
+	}
+	hasConfigGroup := false
+	for i, group := range groups.Groups {
+		if group.Name == configGroup {
+			hasConfigGroup = true
+			hasV1Version := false
+			for _, version := range group.Versions {
+				if version.Version == configVersion {
+					hasV1Version = true
+					break
+				}
+			}
+			if !hasV1Version {
+				groups.Groups[i].Versions = append(groups.Groups[i].Versions, configGroupVersionForDiscovery())
+			}
+			break
+		}
+	}
+	if !hasConfigGroup {
+		groups.Groups = append(groups.Groups, *configAPIGroup())
+	}
+	return groups, nil
+}
+
+func (c *ConfigV1DiscoveryClientShim) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	if groupVersion != configGroupVersion {
+		return c.configClient.Discovery().ServerResourcesForGroupVersion(groupVersion)
+	}
+	fakeList, err := c.fakeClient.Discovery().ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return nil, err
+	}
+	realList, err := c.configClient.Discovery().ServerResourcesForGroupVersion(groupVersion)
+	if err == nil {
+		return newAPIResourceList(groupVersion, fakeList.APIResources, realList.APIResources), nil
+	}
+	if errors.IsNotFound(err) {
+		return newAPIResourceList(groupVersion, []metav1.APIResource{}, fakeList.APIResources), nil
+	}
+	return realList, err
+}
+
+func (c *ConfigV1DiscoveryClientShim) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	groups, resources, err := c.configClient.Discovery().ServerGroupsAndResources()
+	if err != nil {
+		return groups, resources, err
+	}
+	if groups != nil {
+		hasConfigGroup := false
+		for i, group := range groups {
+			if group.Name == configGroup {
+				hasConfigGroup = true
+				hasV1Version := false
+				for _, version := range group.Versions {
+					if version.Version == configVersion {
+						hasV1Version = true
+						break
+					}
+				}
+				if !hasV1Version {
+					groups[i].Versions = append(groups[i].Versions, configGroupVersionForDiscovery())
+				}
+				break
+			}
+		}
+		if !hasConfigGroup {
+			groups = append(groups, configAPIGroup())
```

Comment:
why is this one pointer? (in the duplicate code it is not)

Reply:
`ServerGroupsAndResources` returns `[]*metav1.APIGroup` whereas `ServerGroups` `*metav1.APIGroupList` (`[]metav1.APIGroup` instead)